tests/std/tests/P2505R5_monadic_functions_for_std_expected/test.cpp

```diff
@@ -0,0 +1,377 @@
+    struct Derived : DependentBase<T> {
+        static constexpr bool test() {
+            return permissive();
+        }
+    };
+} // namespace detail
+
+constexpr bool is_permissive = detail::Derived<int>::test();
+
+enum class IsNothrowConstructible : bool { Not, Yes };
+enum class IsNothrowConvertible : bool { Not, Yes };
+
+template <class E>
+[[nodiscard]] constexpr bool IsYes(const E e) noexcept {
+    return e == E::Yes;
+}
+
+struct convertible {
+    constexpr convertible(const int val) noexcept : _val(val) {}
+
+    [[nodiscard]] constexpr bool operator==(const int other) const noexcept {
+        return other == _val;
+    }
+
+    int _val = 0;
+};
+
+struct Immovable {
+    constexpr Immovable(const int x) : v(x) {}
+    Immovable(const Immovable&)            = delete;
+    Immovable(Immovable&&)                 = delete;
+    Immovable& operator=(const Immovable&) = delete;
+    Immovable& operator=(Immovable&&)      = delete;
+    constexpr ~Immovable()                 = default;
+
+    int v;
+};
+
+struct Thingy {
+    expected<int, int> x;
+    constexpr int member_func() const {
+        return 66;
+    }
+};
+
+template <class Expected>
+constexpr void test_impl(Expected&& engaged, Expected&& unengaged) {
+    assert(engaged.has_value());
+    assert(!unengaged.has_value());
+    static_assert(is_same_v<typename remove_cvref_t<Expected>::error_type, int>);
+    using Val = typename remove_cvref_t<Expected>::value_type;
+
+    const auto succeed = [](auto...) { return expected<int, int>{33}; };
+    const auto fail    = [](auto...) { return expected<int, int>(unexpect, 44); };
+
+    {
+        decltype(auto) result = forward<Expected>(engaged).and_then(succeed);
+        static_assert(is_same_v<decltype(result), expected<int, int>>);
+        assert(result == 33);
+    }
+    {
+        decltype(auto) result = forward<Expected>(unengaged).and_then(succeed);
+        static_assert(is_same_v<decltype(result), expected<int, int>>);
+        assert(!result);
+        assert(result.error() == 22);
+    }
+    {
+        decltype(auto) result = forward<Expected>(engaged).and_then(fail);
+        static_assert(is_same_v<decltype(result), expected<int, int>>);
+        assert(!result);
+        assert(result.error() == 44);
+    }
+    {
+        decltype(auto) result = forward<Expected>(unengaged).and_then(fail);
+        static_assert(is_same_v<decltype(result), expected<int, int>>);
+        assert(!result);
+        assert(result.error() == 22);
+    }
+
+    if constexpr (!is_void_v<Val>) {
+        {
+            decltype(auto) result = forward<Expected>(engaged).and_then(&Thingy::x);
+            static_assert(is_same_v<decltype(result), expected<int, int>>);
+            assert(result == 11);
+        }
+        {
+            decltype(auto) result = forward<Expected>(unengaged).and_then(&Thingy::x);
+            static_assert(is_same_v<decltype(result), expected<int, int>>);
+            assert(result.error() == 22);

```

Comment:
It seems that we should first `assert(!result);` before inspecting `result.error()`, for consistency with the test cases above. Occurs below.